File: fluss-rust/bindings/elixir/lib/fluss/table.ex
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

defmodule Fluss.Table do
  @moduledoc """
  A handle to a Fluss table, used to create writers and scanners.
  """

  alias Fluss.Native

  @type t :: reference()

  @spec get(Fluss.Connection.t(), String.t(), String.t()) ::
          {:ok, t()} | {:error, Fluss.Error.t()}
  def get(conn, database, table) do
    conn
    |> Native.table_get(database, table)
    |> Native.await_nif()
  end

  @spec get!(Fluss.Connection.t(), String.t(), String.t()) :: t()
  def get!(conn, database, table) do
    case get(conn, database, table) do
      {:ok, t} -> t
      {:error, %Fluss.Error{} = err} -> raise err
    end
  end

  @spec has_primary_key?(t()) :: boolean()
  def has_primary_key?(table), do: Native.table_has_primary_key(table)

  @spec column_names(t()) :: [String.t()]
  def column_names(table), do: Native.table_column_names(table)
end

File: fluss-rust/bindings/elixir/native/fluss_nif/src/table.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use crate::async_nif;
use crate::connection::ConnectionResource;
use fluss::client::{FlussConnection, FlussTable, Metadata};
use fluss::error::Error;
use fluss::metadata::{Column, TableInfo, TablePath};
use rustler::{Env, ResourceArc, Term};
use std::sync::Arc;

/// Holds the data needed to reconstruct FlussTable (which has a lifetime
/// tied to FlussConnection). We store the Arc<FlussConnection> to keep
/// it alive and reconstruct short-lived FlussTable instances on demand.
pub struct TableResource {
    pub connection: Arc<FlussConnection>,
    pub metadata: Arc<Metadata>,
    pub table_info: TableInfo,
}

impl std::panic::RefUnwindSafe for TableResource {}

#[rustler::resource_impl]
impl rustler::Resource for TableResource {}

impl TableResource {
    pub fn columns(&self) -> &[Column] {
        self.table_info.schema.columns()
    }

    pub fn with_table<T>(&self, f: impl FnOnce(&FlussTable<'_>) -> T) -> T {
        let table = FlussTable::new(
            &self.connection,
            self.metadata.clone(),
            self.table_info.clone(),
        );
        f(&table)
    }
}

#[rustler::nif]
fn table_get<'a>(
    env: Env<'a>,
    conn: ResourceArc<ConnectionResource>,
    database_name: String,
    table_name: String,
) -> Term<'a> {
    let conn_arc = conn.inner.clone();
    async_nif::spawn_task_with_result(env, async move {
        let path = TablePath::new(&database_name, &table_name);
        let (metadata, table_info) = {
            let table = conn_arc.get_table(&path).await?;
            (table.metadata().clone(), table.get_table_info().clone())
        };
        Ok::<_, Error>(ResourceArc::new(TableResource {
            connection: conn_arc,
            metadata,
            table_info,
        }))
    })
}

#[rustler::nif]
fn table_has_primary_key(table: ResourceArc<TableResource>) -> bool {
    table.table_info.has_primary_key()
}

#[rustler::nif]
fn table_column_names(table: ResourceArc<TableResource>) -> Vec<String> {
    table
        .columns()
        .iter()
        .map(|c| c.name().to_string())
        .collect()
}
